pkg/cli/jobctl/commands/run.go

```diff
@@ -0,0 +1,305 @@
+		if err != nil {
+			return 0, err
+		}
+		return m * 1000 * 1000, nil
+	case strings.HasSuffix(limit, "G"):
+		g, err := strconv.ParseInt(limit[:len(limit)-1], 10, 64)
+		if err != nil {
+			return 0, err
+		}
+		return g * 1000 * 1000 * 1000, nil
+	case strings.HasSuffix(limit, "Ki"):
+		ki, err := strconv.ParseInt(limit[:len(limit)-2], 10, 64)
+		if err != nil {
+			return 0, err
+		}
+		return ki * 1024, nil
+	case strings.HasSuffix(limit, "Mi"):
+		mi, err := strconv.ParseInt(limit[:len(limit)-2], 10, 64)
+		if err != nil {
+			return 0, err
+		}
+		return mi * 1024 * 1024, nil
+	case strings.HasSuffix(limit, "Gi"):
+		gi, err := strconv.ParseInt(limit[:len(limit)-2], 10, 64)
+		if err != nil {
+			return 0, err
+		}
+		return gi * 1024 * 1024 * 1024, nil
+	default:
+		return 0, fmt.Errorf("unknown memory limit format: %s (expecting k, M, G, Ki, Mi, Gi suffix)", limit)
+	}
+}
+
+func parseIoLimits(readBps, writeBps, readIops, writeIops []string) ([]*jobv1.IODeviceLimits, error) {
+	// valid formats:
+	// - major:minor=limit
+	// - path=limit
+
+	devices := make(map[string]*jobv1.IOLimits)
+	for _, bps := range readBps {
+		dev, lim, err := parseIoLimit(bps)
+		if err != nil {
+			return nil, fmt.Errorf("invalid io limit format %q: %w", bps, err)
+		}
+		devices[dev] = &jobv1.IOLimits{ReadBps: &lim}
+	}
+	for _, bps := range writeBps {
+		dev, lim, err := parseIoLimit(bps)
+		if err != nil {
+			return nil, fmt.Errorf("invalid io limit format %q: %w", bps, err)
+		}
+		if l, ok := devices[dev]; ok {
+			l.WriteBps = &lim
+		} else {
+			devices[dev] = &jobv1.IOLimits{WriteBps: &lim}
+		}
+	}
+	for _, iops := range readIops {
+		dev, lim, err := parseIoLimit(iops)
+		if err != nil {
+			return nil, fmt.Errorf("invalid io limit format %q: %w", iops, err)
+		}
+		if l, ok := devices[dev]; ok {
+			l.ReadIops = &lim
+		} else {
+			devices[dev] = &jobv1.IOLimits{ReadIops: &lim}
+		}
+	}
+	for _, iops := range writeIops {
+		dev, lim, err := parseIoLimit(iops)
+		if err != nil {
+			return nil, fmt.Errorf("invalid io limit format %q: %w", iops, err)
+		}
+		if l, ok := devices[dev]; ok {
+			l.WriteIops = &lim
+		} else {
+			devices[dev] = &jobv1.IOLimits{WriteIops: &lim}
+		}
+	}
+
+	var limits []*jobv1.IODeviceLimits
+	for dev, l := range devices {
+		limits = append(limits, &jobv1.IODeviceLimits{
+			Device: dev,
+			Limits: l,
+		})
+	}
+	slices.SortFunc(limits, func(a, b *jobv1.IODeviceLimits) int {
+		return cmp.Compare(a.Device, b.Device)
+	})
+	return limits, nil
+}
+
+func parseIoLimit(limit string) (string, int64, error) {
+	dev, lim, ok := strings.Cut(limit, "=")
+	if !ok {
+		return "", 0, fmt.Errorf("expecting 'device=limit'")
+	}
+	limInt, err := strconv.ParseInt(lim, 10, 64)
+	if err != nil {
+		return "", 0, fmt.Errorf("invalid limit value: %w", err)
+	}
+
+	if filepath.IsAbs(dev) {
+		return dev, limInt, nil
+	}
+
+	idLink := filepath.Join("/dev/block/", dev)
+	path, err := os.Readlink(idLink)
```

Comment:
Why is the client application following symlinks in the local system to decide whether or not an IO limit is valid?

Reply:
Fixed